x/proof/keeper/proof_validation.go

```diff
@@ -175,10 +154,87 @@ func (k Keeper) EnsureValidProof(
 		relayBz,
 		serviceRelayDifficulty.GetTargetHash(),
 	); err != nil {
-		return types.ErrProofInvalidRelayDifficulty.Wrapf("failed to validate relay difficulty for service %s due to: %v", sessionHeader.ServiceId, err)
+		return nil, types.ErrProofInvalidRelayDifficulty.Wrapf("failed to validate relay difficulty for service %s due to: %v", sessionHeader.ServiceId, err)
 	}
 	logger.Debug("successfully validated relay mining difficulty")
 
+	// Retrieve the corresponding claim for the proof submitted
+	claim, err := k.queryAndValidateClaimForProof(ctx, sessionHeader, supplierOperatorAddr)
+	if err != nil {
+		return nil, err
+	}
+	logger.Debug("successfully retrieved and validated claim")
+
+	return claim, nil
+}
+
+// EnsureValidProofSignaturesAndClosestPath ensures that the proof submitted by
+// the supplier has a valid relay request/response signatures and closest path
+// with respect to an onchain claim.
+//
+// This function should be called in the EndBlocker corresponding to the block height
+// at which the proof is submitted rather than during proof submission (i.e. SubmitProof).
+//
+// NOTE: A fully valid proof must pass both EnsureWellFormedProof and
+// EnsureValidProofSignaturesAndClosestPath.
+func (k Keeper) EnsureValidProofSignaturesAndClosestPath(
+	ctx context.Context,
+	proof *types.Proof,
+) error {
+	// Telemetry: measure execution time.
+	defer cosmostelemetry.MeasureSince(cosmostelemetry.Now(), telemetry.MetricNameKeys("proof", "validation")...)
+
+	logger := k.Logger().With("method", "EnsureValidProofSignaturesAndClosestPath")
+
+	// Retrieve the supplier operator's public key.
+	supplierOperatorAddr := proof.SupplierOperatorAddress
+	supplierOperatorPubKey, err := k.accountQuerier.GetPubKeyFromAddress(ctx, supplierOperatorAddr)
+	if err != nil {
+		return err
+	}
+
+	sessionHeader := proof.GetSessionHeader()
+
+	// Unmarshal the closest merkle proof from the message.
+	sparseCompactMerkleClosestProof := &smt.SparseCompactMerkleClosestProof{}
+	if err = sparseCompactMerkleClosestProof.Unmarshal(proof.ClosestMerkleProof); err != nil {
+		return types.ErrProofInvalidProof.Wrapf(
+			"failed to unmarshal closest merkle proof: %s",
+			err,
+		)
+	}
+
+	// SparseCompactMerkeClosestProof does not implement GetValueHash, so we need to decompact it.
```

Comment:
```suggestion
	// SparseCompactMerkeClosestProof was intentionally compacted to reduce its onchain state size
	// so it must be decompacted rather than just retrieving the value via GetValueHash (not implemented).
```